api/history.go

```diff
@@ -0,0 +1,102 @@
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+
+	"github.com/gorilla/mux"
+	"github.com/puppetlabs/wash/activity"
+	log "github.com/sirupsen/logrus"
+)
+
+// swagger:response
+//nolint:deadcode,unused
+type journals struct {
+	// in: body
+	Journals []activity.Journal
+}
+
+// swagger:route GET /history history retrieveHistory
+//
+// Get command history
+//
+// Get a list of commands that have been run via 'wash' and when they were run.
+//
+//     Produces:
+//     - application/json
+//
+//     Schemes: http
+//
+//     Responses:
+//       200: journals
+//       400: errorResp
+//       404: errorResp
+//       500: errorResp
+var historyHandler handler = func(w http.ResponseWriter, r *http.Request) *errorResponse {
+	history := activity.History()
+
+	w.WriteHeader(http.StatusOK)
+
+	commands := make([]string, len(history))
+	for i, item := range history {
+		// TODO: send structured data instead of a pre-formatted string
+		commands[i] = item.Start.Format("2006-01-02 15:04") + "  " + item.Description
+	}
+	jsonEncoder := json.NewEncoder(w)
+	if err := jsonEncoder.Encode(&commands); err != nil {
+		log.Warnf("Unable to encode history: %v", err)
+		return unknownErrorResponse(fmt.Errorf("Could not marshal %v: %v", history, err))
+	}
+	return nil
+}
+
+// swagger:route GET /history/{id} journal getJournal
+//
+// Get logs for a particular entry in history
+//
+// Get the logs related to a particular command run via 'wash', requested by
+// index within its activity history.
+//
+//     Produces:
+//     - application/json
+//     - application/octet-stream
+//
+//     Schemes: http
+//
+//     Responses:
+//       200: octetResponse
+//       400: errorResp
+//       404: errorResp
+//       500: errorResp
+var journalHandler handler = func(w http.ResponseWriter, r *http.Request) *errorResponse {
```

Comment:
Hmm, I still like `historyEntryHandler` because it's more RESTy (from what I've seen, GET `/<collection>` is an array of items, and GET `/<collection>/<item_id>` returns more detailed info for a single item, while here, we go from thinking about history entries (collection) to a journal (item), which is weird).

I think the response for `/history/<ix>` could be JSON, maybe something like
```
startTime
journal: {
  id: <id>
  description: <description>
  path: <journal file path>
  content: <content>
}
```

^ We could probably change the `/history` endpoint to return an array of those objects, but omit the content.

Reply:
Might also be worth adding a separate `journal/<id>` endpoint.

Reply:
I eventually want `/history/<id>` to be able to stream updates. I feel like what you described is a more specialized thing.

Reply:
Ok, that makes sense. Can we change the handler's name to `historyEntryHandler` (or something like it) so that it is at least more REST-y? `historyHandler` and `journalHandler` is still confusing to think about.

Reply:
Done.